File: api/v2/model/calendar_event.go
package model

import (
	"app/util"
	"time"

	ics "github.com/arran4/golang-ical"
)

type CalendarEvent struct {
	Id               int32 `gorm:"primaryKey" json:"id"`
	AllDay           bool
	Description      string
	EndAt            string
	StartAt          string
	LastModifiedUser string
	Latitude         *float64
	Longitude        *float64
	Location         *string
	Summary          string
	CalendarDetailId string
	DiscardedAt      *string `gorm:"index"`
	CreatedAt        string
	UpdatedAt        string
}

func (e *CalendarEvent) startAt() time.Time {
	return util.TimeParse(e.StartAt)
}

func (e *CalendarEvent) endAt() time.Time {
	return util.TimeParse(e.EndAt)
}

func (e *CalendarEvent) createdAt() time.Time {
	return util.TimeParse(e.CreatedAt)
}

func (e *CalendarEvent) updatedAt() time.Time {
	return util.TimeParse(e.UpdatedAt)
}

func (e *CalendarEvent) Register(cal *ics.Calendar) {
	u := util.GenUUID()
	event := cal.AddEvent(u)
	event.SetCreatedTime(e.createdAt())
	event.SetStartAt(e.startAt())
	event.SetEndAt(e.endAt())
	event.SetSummary(e.Summary)
	event.SetLocation(*e.Location)
	event.SetDescription(e.Description)
	event.SetGeo(e.Latitude, e.Longitude)
	event.SetClass("PUBLIC")
}

File: api/v2/tests/calendar_event.rs
use api::domain::calendar_event::CalendarEvent;
use chrono::Utc;

#[test]
fn test_new_calendar_event() {
    let event = CalendarEvent::new(
        1,
        String::from("Test Event"),
        String::from("Test Description"),
        String::from("Test Location"),
        Utc::now(),
        Utc::now(),
        0.0,
        0.0,
        false,
    );
    assert_eq!(event.id, 1);
    assert_eq!(event.summary, "Test Event");
    assert_eq!(event.description, "Test Description");
    assert_eq!(event.location, "Test Location");
    assert_eq!(event.latitute, 0.0);
    assert_eq!(event.longitude, 0.0);
    assert_eq!(event.all_day, false);
    assert!(event.created_at <= Utc::now());
    assert!(event.updated_at <= Utc::now());
}

fn test_geo() {
    let latitude = 135.0;
    let longitude = 66.0;
    let event = CalendarEvent::new(
        1,
        String::from("Test Event"),
        String::from("Test Description"),
        String::from("Test Location"),
        Utc::now(),
        Utc::now(),
        latitude,
        longitude,
        false,
    );
    let expects = format!("{};{}", latitude, longitude);
    assert_eq!(event.geo(), expects);
}
